app-misc/pip3line/pip3line-3.4.2.ebuild

```diff
@@ -0,0 +1,82 @@
+# Copyright 1999-2018 Gentoo Authors
+# Distributed under the terms of the GNU General Public License v2
+
+EAPI=6
+
+PYTHON_COMPAT=( python{2_7,3_{4,5,6,7}} )
+
+inherit cmake-utils python-r1
```

Comment:
As @juippis wrote already, you should use `python-single-r1` because your app doesn't build against all selected python impls, but one. `python-r1` isn't the proper way to ask the user to choose one and only one python impl to build against.

Reply:
The intent is to build against two, Python2.7 and Python 3.x, all the time.

Reply:
Huh, really? In the same executable? I was under the impression that multiple python impls could hardly coexist in the same executable. I'm curious about how it's done. Looking...

Reply:
I have been doing this very thing for the past couple of years.

It works and yes I can execute both python 2.7 scripts and Python 3 scripts at the same time at runtime. They are both loaded as external plugin.

There is an issue on Mac platforms though, where only one version can be enabled at the same time. I haven't really investigate this too much.

Reply:
Yeah, I've been looking at it. It looks nicely done. Then sorry, I misled you because I didn't understand that two python impls could be loaded at once. I don't even think that my `python_is_python3` suggestion works. Your previous solution probably worked better. `python-single-r1` isn't the proper eclass to use then.

Things get murky here, however because users could choose 3 python impls and then things start falling apart. I don't think that there are many other examples of dual-python-impls programs in the tree, so to my knowledge, we don't have an elegant mechanism to deal with that situation.

Also, you will want to fix your ebuild to work with python2_7 because `configure` fails when `python2_7` is enabled. It seems that cmake has to be fed a path to it.

Reply:
ok, so moving back to using "${PYTHON_TARGETS}" == *python3*.

Originally I had a separate use flag for both python version (without using python-r1), because I was not sure how the portage system would manage a dual implementation. It's seems to work though.

I don't understand what you mean when you say "you will want to fix your ebuild to work with python2_7"

Do you mean only python2_7 ? not both ?

It compiles fine with both. Will try with only Python2.7.

Reply:
oh right, I get it, the cmake scripts output a warning for Python2.7.

That's, uh, "expected": cmake cannot handle dual Python package search at the same time, so the Python 2.7 version is just using hard-coded paths to solve that, without using the usual find_package(PythonLibs x.x).

As you noted earlier I am probably the only one to build an executable with both Python runtimes. So I am right on the edge case.

Reply:
I have `PYTHON_TARGETS="python2_7 python3_6"` and I get:

```
-- Configuring the Python 2.7 plugin --
CMake Warning at defaultplugins/python27plugin/CMakeLists.txt:45 (message):
  Python 2.7 paths haven't been found or supplied, trying hardcoded values


CMake Error at defaultplugins/python27plugin/CMakeLists.txt:58 (message):
  No Python 2.7 libraries found, Even hardcoded ones
```

builds fine with only `python3_6` though.

Reply:
oh ? that's new.

Reply:
Hum, you... might have to return to your old flags then :) To ensure that proper impls are selected, we'll need a mix of `python/python3` USE flags along with `REQUIRED_USE` rules that ensure that only one impl of each major python version is selected. And then, at configure time we'll need some more work to pass proper python impls to cmake because if we let cmake auto-detect python impls, it might not choose the same one as the one selected through `PYTHON_TARGETS`. Lots of messy work ahead, but it should be doable. You'll need to read through python-related eclasses to figure out what to do.

It's an interesting case. I don't have the time right now, but I'll soon run a little search in the gentoo tree to check if we have a similar situation.

Reply:
I am not sure what to say here, mine works fine:

app-misc/pip3line-3.4.2::local  USE="distorm openssl qscintilla" PYTHON_TARGETS="python2_7 python3_6 ..."

and even with only Python2_7.

Reply:
Oh, I think I know: I'm on the 17.1 profile which removes the `/usr/lib -> /usr/lib64` symlink. So on your machine, your hardcoded lookup of `/usr/lib` works, but only by chance. It should really look in `/usr/$(get_libdir)`.

Reply:
Ah .... that's actually a much bigger issue ...

Reply:
ok, so let's go back to the old flags. T_T

Regarding the cmake script, I have been using it for quite a while now ( > 4 years) and it worked well with python3.x.
The only issue with Python2.7 is that it was selecting python3.x all the time instead of 2.7, so I had to hardcode it (it is working fine on its own though, the issue is only triggered when both modules are enabled simultaneously)

Reply:
I don't think you understand what I mean: we can't let cmake autodetect python. For example, if I have python3.6 and python 3.7 installed on my system and that I build pip3line with `PYTHON_TARGETS="python3_7"`, pip3line's cmake is still going to select python 3.6.

Reply:
Following up on "I'll run a search", I think I found `sys-auth/sssd` which seems to link to python in a similar manner.

Reply:
That's not supposed to happen. I did not have that issue until today, event with multiple minor version of Python 3.

Reply:
Not supposed to happen? It happens, I've tried it. cmake selected python3.6 when I asked it to build with python 3_7.

Reply:
It seems to work for me:

~~~
-- OpenSSL library found ... OK
-- Configuring the Python 3 plugin --
-- Found PythonLibs: /usr/lib/libpython3.7m.so (found suitable version "3.7.0", minimum required is "3") 
-- Configuring the Python 2.7 plugin --
~~~

profile: default/linux/amd64/17.0/desktop/plasma/systemd
cmake : dev-util/cmake-3.9.6::gentoo

Reply:
Also it ignores entirely the defined PYTHON_TARGETS, because cmake does not care about it. Unless there is a way to specify the includes/libs  to cmake to use the highest defined PYTHON_TARGETS. Maybe I should look at the doc ...

Reply:
Your newest commit still has the problem described above. I know, its your first PR and that the review process has been heavy so far, but that program is also a bit special. It needs an extra effort. Sloppiness isn't acceptable. You could also start with a PR for something that's a bit easier to package :)

Read python-r1.eclass carefully, especially `python_setup()` (grep for it in the gentoo tree, many interesting examples), it will help you come up with a solution. You will also need to patch your build system so that it parametrizes python impls to use instead of using its own "best found".

Reply:
(I don't have anything else to package ... :D )

Fair comment, and in addition I haven't looked at the cmake scripts in a while and a lot of garbage had accumulated. I spend some time cleaning them up, but at this point submitting a patch and releasing a brand new cmake script would be the same.

The python cmake scripts can already take paths for include and libs files that would overwrite the usual cmake find. So I guess I need to look closer at this python-r1.eclass now...

Reply:
updated